File: 2022/11.03/황지원/연속 부분 수열 합의 개수.java
import java.util.*;

class Solution {
  	static ArrayList<Integer> list = new ArrayList<>();
	static int size;
	static int arr[];
	static HashSet<Integer> set;

    public int solution(int[] elements) {
        size = elements.length;
		arr = new int[size * 2];
		set = new HashSet<>();
        for (int i = 0; i < size * 2; i++) {
			arr[i] = elements[i % size];
		}
		for (int start = 0; start < size; start++) { // 시작점 잡고
            int sum = 0;
            int i = 0;
            while(i < size && i + start < size * 2) { // 1개부터 size개까지
                sum += arr[i + start];
                set.add(sum);
                i++;
            }
        }	
		return set.size();
	}

}

File: 11.03/조수연/연속 부분 수열 합의 개수.kt
package com.ssafy.algorithm

fun main(){
    val elements = intArrayOf(7,9,1,1,4)
    println(Solution().solution(elements))
}

class Solution {

    lateinit var set : HashSet<Int> //숫자합 set
    lateinit var visited : BooleanArray
    var isComplete = false //다음 인덱스 넘어가는 체크 배열

    fun solution(elements: IntArray): Int {
        set = HashSet()
        visited = BooleanArray(elements.size)

        for (i in 1..elements.size){
            isComplete = false
            combination(0,0,0,i,0,elements)
        }

        return set.size
    }

    private fun combination(size : Int, start : Int,lastIndex : Int , max : Int, sum : Int, elements: IntArray){

        if (size == max){
            set.add(sum)
            isComplete = true
            return
        }

        //원형 수열이므로 마지막인덱스에서도 + size만큼 숫자 뽑기
        for (i in start until elements.size-1+max){
            if (size == 0) isComplete = false //연속된 수열이 아니면 break
            if (size == 0 && i == elements.size) break //마지막인덱스를 지나면 break

            if (!isComplete) combination(size+1,i+1,i,max,sum+elements[i%elements.size],elements)
            else break
        }
    }
}







